Userland/Libraries/LibDeviceTree/DeviceTree.h

```diff
@@ -0,0 +1,191 @@
+            auto sub_property = DeviceTreeProperty { raw_data.slice(offset, sizeof(T)) };
+            auto result = function(sub_property.as<T>());
+            if (result.is_error())
+                return result;
+            if (result.value() == IterationDecision::Break)
+                break;
+        }
+        return {};
+    }
+
+    FixedMemoryStream as_stream() const { return FixedMemoryStream { raw_data }; }
+};
+
+class DeviceTreeNodeView {
+public:
+    bool has_property(StringView prop) const { return m_properties.contains(prop); }
+    bool has_child(StringView prop) const { return m_children.contains(prop); }
+    bool has(StringView prop) const { return has_property(prop) || has_child(prop); }
+
+    Optional<DeviceTreeProperty> get_property(StringView prop) const { return m_properties.get(prop); }
+
+    // FIXME: The spec says that @address parts of the name should be ignored when looking up nodes
+    //        when they do not appear in the queried name, and all nodes with the same name should be returned
+    Optional<DeviceTreeNodeView const&> get_child(StringView name) const { return m_children.get(name); }
+
+    HashMap<StringView, DeviceTreeNodeView> const& children() const { return m_children; }
+    HashMap<StringView, DeviceTreeProperty> const& properties() const { return m_properties; }
+
+    DeviceTreeNodeView const* parent() const { return m_parent; }
+
+    // FIXME: Add convenience functions for common properties like "reg" and "compatible"
+    // Note: The "reg" property is a list of address and size pairs, but the address is not always a u32 or u64
+    //       In pci devices the #address-size is 3 cells: (phys.lo phys.mid phys.hi)
+    //       with the following format:
+    //       phys.lo, phys.mid: 64-bit Address - BigEndian
+    //       phys.hi: relocatable(1), prefetchable(1), aliased(1), 000(3), space type(2), bus number(8), device number(5), function number(3), register number(8) - BigEndian
+
+    // FIXME: Stringify?
+    // FIXME: Flatten?
+    // Note: That we dont have a oder of children and properties in this view
+protected:
+    friend class DeviceTree;
+    DeviceTreeNodeView(DeviceTreeNodeView* parent)
+        : m_parent(parent)
+    {
+    }
+    HashMap<StringView, DeviceTreeNodeView>& children() { return m_children; }
+    HashMap<StringView, DeviceTreeProperty>& properties() { return m_properties; }
+    DeviceTreeNodeView* parent() { return m_parent; }
+
+private:
+    DeviceTreeNodeView* m_parent;
+    HashMap<StringView, DeviceTreeNodeView> m_children;
+    HashMap<StringView, DeviceTreeProperty> m_properties;
+};
+
+class DeviceTree : public DeviceTreeNodeView {
+public:
+    static ErrorOr<NonnullOwnPtr<DeviceTree>> parse(ReadonlyBytes);
+
+    DeviceTreeNodeView const* resolve_node(StringView path) const
+    {
+        // FIXME: May children of aliases be referenced?
+        if (auto alias_list = get_child("aliases"sv); alias_list.has_value()) {
```

Comment:
Aliases are not allowed to contain a `/`, so you should only need to check for aliases if the first char is not a `/`.

Reply:
Interesting, will do that